substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/throttling/JfrAdaptiveSampler.java

```diff
@@ -0,0 +1,201 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr.throttling;
+
+import static com.oracle.svm.core.Uninterruptible.CALLED_FROM_UNINTERRUPTIBLE_CODE;
+import static java.lang.Math.log;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.jdk.UninterruptibleUtils;
+import com.oracle.svm.core.jfr.utils.JfrRandom;
+import com.oracle.svm.core.thread.JavaSpinLockUtils;
+import com.oracle.svm.core.util.TimeUtils;
+
+import jdk.internal.misc.Unsafe;
+
+/**
+ * This class is based on the JDK 23+8 version of the HotSpot class {@code JfrAdaptiveSampler} (see
+ * hotspot/share/jfr/support/jfrAdaptiveSampler.hpp).
+ */
+abstract class JfrAdaptiveSampler {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    protected static final long LOCK_OFFSET = U.objectFieldOffset(JfrAdaptiveSampler.class, "lock");
+    private static final long ACTIVE_WINDOW_OFFSET = U.objectFieldOffset(JfrAdaptiveSampler.class, "activeWindow");
+
+    private final JfrRandom prng;
+    private final JfrSamplerWindow window0;
+    private final JfrSamplerWindow window1;
+
+    @SuppressWarnings("unused") private volatile int lock;
+    protected JfrSamplerWindow activeWindow;
+    protected double avgPopulationSize;
+    private double ewmaPopulationSizeAlpha;
+    private long accumulatedDebtCarryLimit;
+    private long accumulatedDebtCarryCount;
+
+    JfrAdaptiveSampler() {
+        prng = new JfrRandom();
+
+        window0 = new JfrSamplerWindow();
+        window1 = new JfrSamplerWindow();
+        activeWindow = window0;
+    }
+
+    @Uninterruptible(reason = CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+    protected boolean sample(long timestampNs) {
+        boolean expired = activeWindow.isExpired(timestampNs);
+        if (expired) {
+            if (JavaSpinLockUtils.tryLock(this, LOCK_OFFSET)) {
+                /* Recheck under lock if the current window is still expired. */
+                if (activeWindow.isExpired(timestampNs)) {
+                    rotate(activeWindow);
+                }
+                JavaSpinLockUtils.unlock(this, LOCK_OFFSET);
+            }
+            return false;
+        }
+
+        return activeWindow.sample();
```

Comment:
hi @christianhaeubl this reverts the read-write lock changes. Doesn't this still have the same problems you were concerned about (https://github.com/oracle/graal/pull/6899#discussion_r1330104194).

A window may expire after a thread has checked expiry status but before it has sampled with respect to it.

Reply:
Yes, I removed the read/write lock because I don't think that this race can cause any relevant problems (that is probably also the reason why HotSpot doesn't guard against it).